substrate/frame/revive/src/vm/evm.rs

```diff
@@ -110,32 +148,148 @@ pub fn call<'a, E: Ext>(bytecode: Bytecode, ext: &'a mut E, inputs: EVMInputs) -
 	let table = instruction_table::<'a, E>();
 	let result = run(&mut interpreter, &table);
 
-	if result.is_error() {
-		Err(Error::<E::T>::ContractTrapped.into())
-	} else {
-		Ok(ExecReturnValue {
-			flags: if result.is_revert() { ReturnFlags::REVERT } else { ReturnFlags::empty() },
-			data: result.output.to_vec(),
+	instruction_result_into_exec_error::<E>(result.result)
+		.map(Err)
+		.unwrap_or_else(|| {
+			Ok(ExecReturnValue {
+				flags: if result.is_revert() { ReturnFlags::REVERT } else { ReturnFlags::empty() },
+				data: result.output.to_vec(),
+			})
 		})
-	}
 }
 
-/// Runs the EVM interpreter
-fn run<WIRE: InterpreterTypes>(
-	interpreter: &mut Interpreter<WIRE>,
-	table: &revm::interpreter::InstructionTable<WIRE, DummyHost>,
+/// Runs the EVM interpreter until it returns an action.
+fn run<'a, E: Ext>(
+	interpreter: &mut Interpreter<EVMInterpreter<'a, E>>,
+	table: &revm::interpreter::InstructionTable<EVMInterpreter<'a, E>, DummyHost>,
 ) -> InterpreterResult {
 	let host = &mut DummyHost {};
-	let action = interpreter.run_plain(table, host);
-	match action {
-		InterpreterAction::Return(result) => return result,
-		InterpreterAction::NewFrame(_) => {
-			// TODO handle new frame
-			InterpreterResult::new(
-				revm::interpreter::InstructionResult::FatalExternalError,
-				Default::default(),
-				interpreter.gas,
-			)
+	loop {
+		let action = interpreter.run_plain(table, host);
+		match action {
+			InterpreterAction::Return(result) => {
+				log::info!("Return {:?}", result);
```

Comment:
use debug and LOG_TARGET

Reply:
can you add that as well

```
			debug_assert!(
				result.gas.limit() == 0 &&
					result.gas.remaining() == 0 &&
					result.gas.refunded() == 0,
				"Interpreter gas state should remain unchanged; found: {:?}",
				result.gas,
			);
```

to make sure we don't use the gas field here

Reply:
why do we need this assert?

Reply:
want to make sure we haven't left any instructions that touches the result.gas that we don't use